packages/foam-core/src/janitor/index.ts

```diff
@@ -0,0 +1,79 @@
+import { Position } from 'unist';
+import dashify from 'dashify';
+import { Note, NoteGraph } from '../index';
+import {
+  createMarkdownReferences,
+  stringifyMarkdownLinkReferenceDefinition,
+} from '../markdown-provider';
+import { getHeadingFromFileName } from '../utils'
+
+export interface TextEdit {
+  range: Position;
+  newText: string;
+}
+
+export const generateLinkReferences = (note: Note, ng: NoteGraph): TextEdit | null => {
+  const newReferences = createMarkdownReferences(ng, note.id).map(
+    stringifyMarkdownLinkReferenceDefinition
+  ).join('\n');
+
+  if (note.definitions.length === 0) {
+    if (newReferences.length === 0) {
+      return null;
+    }
+
+    const padding = note.end.column === 1 ? note.eol : `${note.eol}${note.eol}`;
+    return {
+      newText: `${padding}${newReferences}`,
+      range: {
+        start: note.end,
+        end: note.end,
+      },
+    };
+  } else {
+    const first = note.definitions[0];
+    const last = note.definitions[note.definitions.length - 1];
+
+    const oldRefrences = note.definitions.map(stringifyMarkdownLinkReferenceDefinition).join(note.eol);
+
+    if (oldRefrences === newReferences) {
+      return null;
+    }
+
+    return {
+      // @todo: do we need to ensure new lines?
+      newText: `${newReferences}`,
+      range: {
+        start: first.position!.start,
+        end: last.position!.end,
+      },
+    };
+  }
+};
+
+export const generateHeading = (note: Note): TextEdit | null => {
+  // Note: This may not work if the heading is same as the file name
+  if (note.title !== note.id) {
```

Comment:
Hey @riccardoferretti!

We wanted to check if the file already has a heading or not.  If `note.id` and `note.title` are the same then this probably means that the note didn't have a heading tag and we generate a new one.

Not sure how else can we do this without traversing the AST again.

Reply:
This relies on `let title = id;` in `createNoteFromMarkdown`. It works, but agree with Riccardo it's a little sketchy. A better solution would be to make `title` nullable and not populate it with id by default. 

That change will likely break something else however, so if we do it, we'll have to understand where the title is used.

Reply:
Oh, I see what you mean.
That would work now, but be aware that you are making a fairly big assumption about both what's in the `note.id` and _how_ the `note.title` is computed. The check is using an implementation detail of `Note`, that is, the reason why you know this check works is because you have seen the code that provides the default value to `title`.

As a workaround you could add a test that would fail in this case, so we can catch it shall we make changes to `Note` (which I think are coming). 

Alternatively we could explore whether we really need `title` to default to `id`, or whether we could leave it `null` if there is no title in the note. This wold also solve the `Note: ` you put in the line above your check.
@jevakallio might also have an idea here.
 
Something to think about :)

Reply:
@jevakallio our comments overlapped, but it seems we agree the `null` route is the right way forward.

Reply:
We needed this PR to get merged quicky. Hence, I have created #131. Will fix it soon :)